### ISSUE TITLE
balance of contrastive training

### ISSUE THREAD

**Issue by lixiang-ucas** [fanq15/FSOD-code#28]
Dear Qi, I have a small question here. In your paper, for the contrastive training part, you write "We keep the ratio as 1:2:1 for the foreground proposal and positive support pairs (pf ; sp), background proposal and positive support pairs (pb; sp), and proposal (foreground or background) and negative support pairs (p; sn)". But in your code, as a pasted below, it seems you are using 1:2:1 for the foreground proposal, background proposal and positive support pairs (pb; sp), background proposal negative support pairs (pb; sn). If I understand correctly, "real_bg_inds < int(rois_label.shape[0] * 0.5" indicates those positive supports and "real_bg_inds < int(rois_label.shape[0] * 0.5" indicates those negative supports, which means, in a 2-way-5-shot setting, the first 5 are positive supports and the remaining 5 are negative supports.

Line 174-190 in './lib/modeling/fast_rcnn_heads.py':
    rois_label = Variable(torch.from_numpy(label_int32.astype('int64'))).cuda(device_id)
    cls_score_softmax = torch.nn.functional.softmax(cls_score_ori, dim=1)
    fg_inds = (rois_label == 1).nonzero().squeeze(-1)
    bg_inds = (rois_label == 0).nonzero().squeeze(-1)
    bg_cls_score_softmax = cls_score_softmax[bg_inds, :]

    bg_num_0 = max(1, min(fg_inds.shape[0] * 2, int(rois_label.shape[0] * 0.25)))
    bg_num_1 = max(1, min(fg_inds.shape[0], bg_num_0))

    sorted, sorted_bg_inds = torch.sort(bg_cls_score_softmax[:, 1], descending=True)

    real_bg_inds = bg_inds[sorted_bg_inds]
    real_bg_topk_inds_0 = real_bg_inds[real_bg_inds < int(rois_label.shape[0] * 0.5)][:bg_num_0]
    real_bg_topk_inds_1 = real_bg_inds[real_bg_inds >= int(rois_label.shape[0] * 0.5)][:bg_num_1]

    topk_inds = torch.cat([fg_inds, real_bg_topk_inds_0, real_bg_topk_inds_1], dim=0)
    loss_cls = F.cross_entropy(cls_score_ori[topk_inds], rois_label[topk_inds])

**Comment by fanq15** [comment]
The `fg_inds.shape[0]` is the number of (foreground proposal, positive support), the `bg_num_0` is for (background proposal, positive support), and the `bg_num_1` is for (proposal, negative support). The `real_bg_inds` is the sorted proposal index which is in [0, 255], where the [0,127] is for the positive support and the [127, 255] is for the negative support. I select the top `bg_num_0` proposals in the first part for the (background proposal, positive support) and select the top `bg_num_1` proposals in the first part for the (proposal, negative support). You can print the output of this code part for better understanding.

**Comment by lixiang-ucas** [comment]
Thanks for your kind reply. I'll try printing these outputs with real data. Right now, I'm still downloading the dataset.
